HDDS-6856. HA aware reads from Snapshots

ivandika3 opened the pull request:
## What changes were proposed in this pull request?

This adds a support for Ozone client to be able to do read-only snapshot operations (GetSnapshotInfo, ListSnapshots, SnapshotDiff, ListSnapshotDiffJobs) from non-leader OM. For example, currently Snapdiff request will always go to the current OM leader. The leader will start a Snapdiff job if it does not exist yet and once it's finished, the user can rerun the snapdiff command again to receive the snapshot diffs returned by the Snapdiff job. However, if there is a OM leader change and another snapshot diff is called, the new OM leader needs to start the Snapdiff job again and the client will receive that the snapdiff is not ready yet, although it was ready at the previous leader.

This adds the following features
- Single (no-failover) proxy implementation to pick a specific OM node to get snapshot from `SingleOMFailoverProxyProviderBase`
- Supports for non-leader OM node to serve snapshot-related reads
  - This is supported by adding a new `omNodeId` proto field which if set means that the OM request is intended for the particular OM node 
- Ozone snapshot shell supports to read snapshots from followers

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6856

## How was this patch tested?

Unit test (proxy provider implementation), integration test (Ozone client), and acceptance test (Ozone shell).

peterxcli commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerProtocolServerSideTranslatorPB.java:
```diff
@@ -219,14 +219,35 @@ private OMResponse submitReadRequestToOM(OMRequest request)
     RaftServerStatus raftServerStatus = omRatisServer.checkLeaderStatus();
-    if (raftServerStatus == LEADER_AND_READY ||
-        request.getCmdType().equals(PrepareStatus)) {
+    if (raftServerStatus == LEADER_AND_READY || request.getCmdType().equals(PrepareStatus) ||
+        (raftServerStatus == NOT_LEADER && allowSnapshotReadFromFollower(request))) {
       return handler.handleReadRequest(request);
     } else {
       throw createLeaderErrorException(raftServerStatus);
     }
   }
 
+  private boolean allowSnapshotReadFromFollower(OMRequest omRequest) {
+    if (!omRequest.hasOmNodeId()) {
+      return false;
+    }
+
+    if (!omRequest.getOmNodeId().equals(ozoneManager.getOMNodeId())) {
+      return false;
+    }
+
+    switch (omRequest.getCmdType()) {
+    case SnapshotDiff:
+    case ListSnapshot:
+    case GetSnapshotInfo:
+    case ListSnapshotDiffJobs:
+    case CancelSnapshotDiff:
+      return true;
+    default:
+      return false;
+    }
+  }
+
```
Should this return an error that is something like "OmNodeId specify, but COMMAND_TYPE is not supported to read from follower"?
```suggestion
    if (raftServerStatus == LEADER_AND_READY || request.getCmdType().equals(PrepareStatus)) {
      return handler.handleReadRequest(request);
    } else if (raftServerStatus == NOT_LEADER && readFromFollower(omRequest))
      // handle follower read
      boolean handle = false
      if (isSnapshotCmd(omRequest)) {
         try {
            allowSnapshotReadFromFollower(request)
         } catch(FollowerReadException) {
           ...
         }
      } else {
          throw ServiceException(new FollowerReadUnsopportedCmdExCeption());
      }
      
      if (handle) return handler.handleReadRequest(request);
    } else {
      throw createLeaderErrorException(raftServerStatus);
    }
  }
  
  private boolean readFromFollower(OMRequest omRequest) {
    return omRequest.hasOmNodeId();
  }

  private boolean allowSnapshotReadFromFollower(OMRequest omRequest) throw SnapshotFollowerReadException {
    if (!omRequest.getOmNodeId().equals(ozoneManager.getOMNodeId())) {
      return false;
    }

    switch (omRequest.getCmdType()) {
    case SnapshotDiff:
    case ListSnapshot:
    case GetSnapshotInfo:
    case ListSnapshotDiffJobs:
    case CancelSnapshotDiff:
      return true;
    default:
      throw SnapshotFollowerUnSupportedCmdException(omRequest.getCmdType());
    }
  }
```
above is just a rough suggestion.

ivandika3 replied:
Thanks for the review. Throwing a special exception requires further handling in the client failover proxy handler, but currently any `OMNotLeaderException` from follower will fail the client immediately (see: https://github.com/apache/ozone/blob/ba65f759a05b975e788d3cb55d06df6e843aed2b/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/SingleOMFailoverProxyProviderBase.java#L174). So the behavior should remain unchanged.

I'll keep this in mind if we need this in the future.

peterxcli replied:
Thanks for the clarification.

Could you help to resolve this conversion, as this is quite long😅

peterxcli commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerProtocolServerSideTranslatorPB.java:
```diff
@@ -219,14 +219,35 @@ private OMResponse submitReadRequestToOM(OMRequest request)
+      return false;
+    }
+
+    if (!omRequest.getOmNodeId().equals(ozoneManager.getOMNodeId())) {
+      return false;
+    }
```
Could be another type of exception like `NodeIdMistach`

ivandika3 replied:
Good point, I have added handling for OM node ID mismatch. By right this shouldn't happen, but this should catch bugs early.

peterxcli reviewed (commented):
Left a few comments, feel free to take a look and see if they make sense. Thanks!

SaketaChalamchala commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:
```diff
@@ -1020,6 +1020,30 @@ public OzoneSnapshot getSnapshotInfo(String volumeName,
+  public OzoneSnapshot getSnapshotInfo(String volumeName,
+                                       String bucketName,
+                                       String snapshotName,
+                                       String omNodeId) throws IOException {
```
Thanks for the patch @ivandika3. Can we have a single snapshot function definition and have clients set null omNodeIds if `--om-node-id` is not specified?

ivandika3 replied:
Thank you for the review. Previously, I have two different APIs to be more explicit to the client that if `omNodeId` is specified, it will be definitely sent to that OM node and not the leader.

However, I think it's better to unify it.

SaketaChalamchala commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneManagerVersion.java:
```diff
@@ -54,6 +54,9 @@ public enum OzoneManagerVersion implements ComponentVersion {
       "OzoneManager version that supports S3 list multipart uploads API with pagination"),
+
+  SNAPSHOT_READ_FROM_NON_LEADER(12, "OzoneManager version that supports " +
+      "snapshot read operations from non-leader"),
```
nit: can you change non-leader to follower everywhere?

ivandika3 replied:
I intentionally use "non-leader" term so that it also includes future OM listener. See https://issues.apache.org/jira/browse/HDDS-11523.

errose28 commented:
Thanks for working on this @ivandika3. I only briefly looked at the code, but can we implement this such that the client/server API required for this change can be reused if/when we add general purpose Ratis based read from follower?

For example, with Ratis read-from-follower, I imagine the implementation would look something like this:
- Client gives request to leader OM with an option that says it is ok with reading from a follower for this request.
- OM leader either services the request itself, or load balances by returning options for other followers which have applied all requests.
- If given a list of other OMs, client will try to have them service its request by contacting them directly.

In this implementation, since we are not yet using Ratis to determine who has the snapshots, the same request flow could be used:
- Client gives request to leader OM with an option that says it is ok with reading from a follower for this request.
  - Leader OM only acknowledges this option for snapshot read requests.
- OM leader returns addresses of both followers for the client to try, since we don't have read-from-follower implemented.
- Client tries each OM returned until it finds one that has the snapshot, and then does the read there.

The only snapshot specific aspect here would be the ability to read snapdiff from a specific OM. I would assume the snapdiff request outputs the host that the job is running on so that the user can feed this back into requests for their snapdiff. I don't think general snapshot list/info CLIs would need to be updated since they could use the same general flow as above. Since snapdiff is already OM specific, we should probably implement this part of the change on its own (to account for leader changes) and then come back to this PR as an enhancement on top of that.

It would be good to avoid adding anything to the client/server code that would just get deprecated later, because it adds complexity to our cross compatibility support. I imagine general read-from-follower would be able to replace a lot of the snapshot specific changes here, so we should probably think in that direction.

Alternatively, given that this change is almost 2k lines long, we could look at the level of effort for supporting general purpose Ratis read-from-follower, and only enabling it for snapshot use cases to start out. I would imagine it is about the same amount of changes or less assuming Ratis has already laid the groundwork to help us out.

ivandika3 commented on hadoop-ozone/cli-shell/src/main/java/org/apache/hadoop/ozone/shell/snapshot/SnapshotDiffHandler.java:
```diff
@@ -91,6 +92,11 @@ public class SnapshotDiffHandler extends Handler {
   private boolean json;
 
+  @CommandLine.Option(
+      names = {"-n", "--om-node-id"},
```
Yes, otherwise it would be sent to the leader.

ivandika3 commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/SingleOMFailoverProxyProviderBase.java:
```diff
@@ -0,0 +1,252 @@
+/**
+ * An implementation of {@link FailoverProxyProvider} which does nothing in the
+ * event of OM failover, and always returns the same proxy object. In case of OM failover,
+ * client will keep retrying to connect to the same OM node.
```
Yes, the client will eventually fail. This proxy provider is specifically used to create a stick to a single OM node to reduce the load on OM leader. 

Additionally, we want to make the snapdiff status predictable (e.g. after the first snapdiff status is DONE, the subsequent snapdiff API call on the same OM node should also return DONE). This also reduce the overall number of snapdiff jobs in different OM nodes.

In the context of Cross-Region Bucket Replication (Incremental Replication), there will be a master process that will detect the OM node failure and will reassign the worker to a run Snapdiff from a new OM node.

ivandika3 commented on hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:
```diff
@@ -179,6 +179,13 @@ message OMRequest {
 
+  // If omNodeID is specified, this means that this client request is specifically
+  // made to the particular OM node, regardless whether the OM node is a leader.
+  // OM needs to rejects the request if the omNodeID is not equal to its own node ID.
```
1. Currently OM node will reject the request since this means that the OM nodes configuration of the client and the server are not synchronized. Not sure if we can add logic like the current suggested address mechanism.
2. Leadership change should not affect the request, the client will still contact the same node, even though it just became a leader. In the context of Cross Region Bucket Replication, the intent is for the client to stick to an OM listener which will never be a leader, but in the future if a worker can listen to a follower, we can add a reassignment logic to pick another OM follower. However, as mentioned in another comment, we want to support predictability of snapdiff and reduction of the number of snapdiff jobs.

ivandika3 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerProtocolServerSideTranslatorPB.java:
```diff
@@ -187,14 +188,50 @@ private OMResponse submitReadRequestToOM(OMRequest request)
+    return true;
+  }
+
+  private boolean allowSnapshotReadFromFollower(OMRequest omRequest) {
```
Currently the purpose is to support reading snapshots and keys under snapshots. We can change this once we support non-snapshot read from follower (e.g. https://github.com/apache/ozone/pull/5288).

ivandika3 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerProtocolServerSideTranslatorPB.java:
```diff
@@ -187,14 +188,50 @@ private OMResponse submitReadRequestToOM(OMRequest request)
     RaftServerStatus raftServerStatus = omRatisServer.checkLeaderStatus();
-    if (raftServerStatus == LEADER_AND_READY ||
-        request.getCmdType().equals(PrepareStatus)) {
+    boolean isNodeIdSpecified = validateNodeId(request);
```
```java
    if (!omRequest.getOmNodeId().equals(ozoneManager.getOMNodeId())) {
      throw new ServiceException(new OMNodeIdMismatchException("OM request node ID is " + omRequest.getOmNodeId() +
          ", but the current OM node ID is " + ozoneManager.getOMNodeId()));
    }
```
`validateNodeId` will throw `OmNodeIdMistmatchException` if the OM request specify the OM Node ID but the current OM Node ID is wrong. This is run before the leadership check, so it won't pass to leader.

Please let me know if I miss something as well.

ivandika3 commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:
```diff
@@ -1009,14 +1009,29 @@ public void deleteSnapshot(String volumeName,
         "bucket can't be null or empty.");
     Preconditions.checkArgument(StringUtils.isNotBlank(snapshotName),
         "snapshot name can't be null or empty.");
+    if (StringUtils.isNotBlank(omNodeId) &&
```
I think we can let this be since the IOException message is different for different snapshot API calls, abstracting to a function also need to pass a string like "snapshot read", "snapshot diff", etc.

I can abstract out the "if condition" if needed.

ivandika3 commented on hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/ozone/Hadoop27RpcTransport.java:
```diff
@@ -101,18 +145,35 @@ public Text getDelegationTokenService() {
+    for (HadoopRpcSingleOMFailoverProxyProvider<OzoneManagerProtocolPB> failoverProxyProvider
+        : omFailoverProxyProviders.values()) {
+      failoverProxyProvider.close();
+    }
```
Thank you for the review. Updated.

ivandika3 commented:
@errose28 Thanks for the initial review.

> Thanks for working on this @ivandika3. I only briefly looked at the code, but can we implement this such that the client/server API required for this change can be reused if/when we add general purpose Ratis based read from follower?

It's possible, but it requires first to consolidate designs for general purposed Ratis based read from follower. Additionally, the use case for this patch is not for general purpose read, but to be able to continuously read from a single OM node, mentioned in the design doc of "Cross-Region Bucket Replication" (https://issues.apache.org/jira/browse/HDDS-12307).

> For example, with Ratis read-from-follower, I imagine the implementation would look something like this:
> 
> * Client gives request to leader OM with an option that says it is ok with reading from a follower for this request.
> * OM leader either services the request itself, or load balances by returning options for other followers which have applied all requests.
> * If given a list of other OMs, client will try to have them service its request by contacting them directly.
> 
> In this implementation, since we are not yet using Ratis to determine who has the snapshots, the same request flow could be used:
> 
> * Client gives request to leader OM with an option that says it is ok with reading from a follower for this request.
>   
>   * Leader OM only acknowledges this option for snapshot read requests.
> * OM leader returns addresses of both followers for the client to try, since we don't have read-from-follower implemented.
> * Client tries each OM returned until it finds one that has the snapshot, and then does the read there.

There have been some previous discussions regarding the read-from-follower
- OM HA: support read from followers (https://issues.apache.org/jira/browse/HDDS-9279 and https://github.com/apache/ozone/pull/5288): This should be the simplest one since there is very small changes needed for the OM (server-side). However, currently our current OM setup does not guarantee linearizability (AFAIK we only support read-after-write consistency). Therefore, when the linearizability read was tested it causes the overall throughput (read + write) to be reduced by quite a lot. Refer to https://docs.google.com/document/d/1xVkaQYDXJmztETJVZQHkkij_j8j6MGQ4XB8ehathhG8/edit?tab=t.0#heading=h.o61uifuxltgn for the benchmark results.
- Allow slightly stale reads: @whbing team implemented readonly client that will periodically probe the OM nodes for the appliedIndex and latency. AFAIK, based on the last probes information, the readonly client will read the "fastest" OM follower (the one with the largest appliedIndex). This might cause stale reads, but this can increase the read throughput. I suppose that this requires additional heartbeat protocols and possibly changing the OM request and response. 
- We can have similar mechanism as HDFS Observer Read (https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/ObserverNameNode.html). AFAIK, if it's applied to Ozone, client will request the lastAppliedIndex of the OM leader (similar to msync call) and then will use the appliedIndex when sending request to the OM follower. This requires adding appliedIndex in the OM request and adding msync protocol. Also we need to change the OM failover proxy provider to allow read from follower.
- I recorded some of my throughs while brainstorming the possible implementation in https://issues.apache.org/jira/browse/HDDS-9279 comment, these are just loose thoughts to highlight the elements to take into account.

Therefore, if we want to not change the protobuf field in the future, we have to be pretty sure about the general purpose read from follower implementation, which requires additional time and effort.

> The only snapshot specific aspect here would be the ability to read snapdiff from a specific OM. I would assume the snapdiff request outputs the host that the job is running on so that the user can feed this back into requests for their snapdiff. I don't think general snapshot list/info CLIs would need to be updated since they could use the same general flow as above. Since snapdiff is already OM specific, we should probably implement this part of the change on its own (to account for leader changes) and then come back to this PR as an enhancement on top of that.

The issue is that snapdiff requires that the two snapshots exists, which might not happen on a slow follower even if the create snapshot has been applied in the leader. Therefore, we need to also support listing snapshots and get snapshot info operation. During the incremental replication in "Cross-Region Bucket Replication", the Syncer will wait until the previous snapshot and new snapshot have been created in the OM listener, and then the Syncer will send a SnapDiff request to OM to get the Snapdiff to replicate to the target bucket.

> It would be good to avoid adding anything to the client/server code that would just get deprecated later, because it adds complexity to our cross compatibility support. I imagine general read-from-follower would be able to replace a lot of the snapshot specific changes here, so we should probably think in that direction.

I agree. If we are still not sure, it might be better to defer this patch in the feature or create a feature branch until we can commit with the client and server changes.

ivandika3 commented:
After discussion with @xichen01 , we identified an issue of using `omNodeId` in the OM request might have issues where the client and server configurations have different OM node IDs.

For example, OM server configuration

```
  <property> 
    <name>ozone.om.service.ids</name>
    <value>service1</value>
  </property>
  <property>
    <name>ozone.om.address.service1.om1</name>
    <value>10.0.0.1</value>
  </property>
  <property>
    <name>ozone.om.address.service1.om2</name>
    <value>10.0.0.2</value>
  </property>
  <property>
    <name>ozone.om.address.service1.om3</name>
    <value>10.0.0.3</value>
  </property>
```

OM client configuration

```
  <property> 
    <name>ozone.om.service.ids</name>
    <value>service1</value>
  </property>
  <property>
    <name>ozone.om.address.service1.om4</name>
    <value>10.0.0.1</value>
  </property>
  <property>
    <name>ozone.om.address.service1.om5</name>
    <value>10.0.0.2</value>
  </property>
  <property>
    <name>ozone.om.address.service1.om6</name>
    <value>10.0.0.3</value>
  </property>
```

The client request will always be rejected since the client OM node ID is always different than the leader.

One possible solution is to have a simple flag to allow read from non-leader instead of omNodeId. If the flag is set, then read from non-leader is allowed, if flag is not set, then read from non-leader is disallowed. We can remove the OM node ID mismatch logic entirely.

Please let me know what you think.

jojochuang commented:
Overall looks good. 
Something like this should come with a written design doc. Here's a little bit of design inferred from the implementation as I went through the code:

(1) the design does not assume consistency -- client is responsible for ensuring the result is consistent. This is largely fine as these operations work for read-only snapshots. Though extra caution should be taken to confirm the snapshots are available.
(2) client explicitly selects a OM to route the requests to. If a failure happens, client retries the same OM.
(3) For requests that involve iterators (listSnapshots, SnapshotDiff) the same OM is used throughout. It does not round robin or bounce over.

The use of isDirect to force follower OM to process read-only snapshot operation is a simple solution to provide read from follower functionality. It may not be useful or would need to be extended in the future to support a fully consistent read from follower model (similar to HDFS NameNode Consistent read from Standby ([HDFS-12943](https://issues.apache.org/jira/browse/HDFS-12943)), or the one that's built on top of linearizable Ratis.

swamirishi commented:
@ivanandikashopee Amongst the snapshot operation GetSnapshotInfo, ListSnapshots, ListSnapshotDiffJobs are relatively lightweight operations and only SnapshotDiff is actually an expensive operation. I don't really like the idea of client dictating which OM should run the snapshotDiff operation. The leader should dictate this. The leader can perform the following operation on each and every snapshot diff operation:

1. Client asks leader to run a snapshot diff operation.
2. Let us add a command like CheckSnapshotDiif command on OM with request and response of the following structure.
```
Request : 
       volumeName 
       bucketName
       sourceSnapshot
       targetSnapshot

Response:
       omEndpoint
       totalNumberOfSnapshotDiffOpsInQueue(optional) // only to be returned when running on follower
       hasOp (true if the snapshot diff is already in progress or results are already computed)
       snapshotsExists(true if both source and target snapshot exist otherwise false)[Snapshots need not have been created on the follower]
```
 This operation when run on leader where the leader figures out if it has already computed snapshot diff in it's own local system if yes then returns it's own endpoint as the response. Otherwise the leader should hit the same command on the follower where it checks if any of the followers has done performed this operation which as a response should also return the total number of snapshotDiff operations in the follower's own queue.
If the snapshot diff is already being computed in some node then the leader can just return the endpoint of the follower. Otherwise based on the totalNumberOfSnapshotDiffOpsInQueue in the responses sent by the followers the leader can just load balance the snapshot diff operation and return the om endpoint where this operation is supposed to run.

3. Now given that the client has the endpoint, it can run the snapshotDiff command on the particular endpoint and can poll the particular om for the snapshot diff result.

ivandika3 commented:
@jojochuang Addressing your comments.

> Something like this should come with a written design doc

Ok, let's keep this patch on hold first since seems there are consensus needed on the overall designs between the community's expectation of general transparent snapshot read API vs explicitly pick a single OM which is created in the context of "Cross-region bucket replication" to ensure predictability of the Syncer logic.

> (1) the design does not assume consistency -- client is responsible for ensuring the result is consistent. This is largely fine as these operations work for read-only snapshots. Though extra caution should be taken to confirm the snapshots are available.

Yes, correct. It requires the client to keep checking whether the snapshot exists in a non-leader OM.

> (2) client explicitly selects a OM to route the requests to. If a failure happens, client retries the same OM.
> (3) For requests that involve iterators (listSnapshots, SnapshotDiff) the same OM is used throughout. It does not round robin or bounce over.

Yes, there is no automatic failover handling. Clients is responsible to catch the exception and retry another OM to pick another OM.

> The use of isDirect to force follower OM to process read-only snapshot operation is a simple solution to provide read from follower functionality. It may not be useful or would need to be extended in the future to support a fully consistent read from follower model (similar to HDFS NameNode Consistent read from Standby ([HDFS-12943](https://issues.apache.org/jira/browse/HDFS-12943)), or the one that's built on top of linearizable Ratis.

Yes, we need to create a consensus on how to approach this.
